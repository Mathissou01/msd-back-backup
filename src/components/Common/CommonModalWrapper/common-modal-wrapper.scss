@use "src/styles/01-settings/settings.colors" as colors;
@use "src/styles/02-tools/tools.shadows" as shadows;

.c-CommonModalWrapper {
  height: 100%;
  overflow: hidden;

  &:not(:empty) {
    position: absolute;
    top: 0;
    z-index: 10;
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100vw;
    height: 100vh;
    padding: 20px 0;
    background-color: rgb(0 0 0 / 50%);
  }

  &__Modal {
    @include shadows.shadow(2);

    position: relative;
    display: flex;
    flex-direction: column;
    padding: 36px;
    background-color: colors.$white;
    border-radius: 10px;

    &_sizeMedium {
      width: 58%;
    }

    &_sizeLarge {
      width: 95%;
    }
  }

  &__Close {
    position: absolute;
    top: 15px;
    right: 15px;
    width: 40px;
    height: 40px;
    background: url("data:image/svg+xml,%3Csvg width='24' height='24' viewBox='0 0 24 24' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='m.75 23.249 22.5-22.5m0 22.5L.75.749' stroke='%23030F40' stroke-width='1.5' stroke-linecap='round' stroke-linejoin='round'/%3E%3C/svg%3E");
    background-repeat: no-repeat;
    background-position: center;
    background-size: 23px;
  }

  hgroup {
    margin-bottom: 30px;
  }
}
