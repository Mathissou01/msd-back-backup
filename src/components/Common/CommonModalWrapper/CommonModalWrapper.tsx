import classNames from "classnames";
import React, {
  forwardRef,
  ReactNode,
  useCallback,
  useEffect,
  useId,
  useImperativeHandle,
  useRef,
  useState,
} from "react";
import { createPortal } from "react-dom";
import { FocusFirstElement } from "../../../lib/utilities";
import "./common-modal-wrapper.scss";

export enum ICommonModalWrapperSize {
  MEDIUM = "medium",
  LARGE = "large",
}

interface ICommonModalWrapperProps {
  onClose?: () => void;
  children: ReactNode;
  size?: ICommonModalWrapperSize;
}

export type CommonModalWrapperRef = {
  toggleModal: (setValue?: boolean, isCancel?: boolean) => void;
};

export default forwardRef(function CommonModalWrapper(
  {
    onClose,
    children,
    size = ICommonModalWrapperSize.MEDIUM,
  }: ICommonModalWrapperProps,
  parentRef,
) {
  const modalId = useId();
  const ref = useRef<Element | null>(null);
  const [isMounted, setMounted] = useState<boolean>(false);

  const handleToggle = useCallback(
    (setValue?: boolean) => {
      ref.current = document.querySelector<HTMLElement>("#modal-portal");
      const newValue = setValue ?? !isMounted;
      setMounted(newValue);
      if (!newValue && onClose) onClose();
    },
    [isMounted, onClose],
  );

  useEffect(() => {
    function handleEscapeKey(event: KeyboardEvent) {
      if (event.code === "Escape") {
        handleToggle(false);
      }
    }

    document.addEventListener("keydown", handleEscapeKey);
    return () => document.removeEventListener("keydown", handleEscapeKey);
  }, [handleToggle]);

  useImperativeHandle(
    parentRef,
    () => ({
      toggleModal(setValue?: boolean) {
        handleToggle(setValue);
      },
    }),
    [handleToggle],
  );

  const focusRef = useCallback((node: HTMLDivElement) => {
    FocusFirstElement(node);
  }, []);

  function handleKeyDown(e: React.KeyboardEvent<HTMLDivElement>) {
    if (e.key !== "Tab" || !ref.current) return;
    const parentElement = ref.current?.querySelector(`[id="${modalId}"]`);
    const focusableModalElements = parentElement?.querySelectorAll(
      "a[href], button:not([disabled]), textarea, input, select",
    );
    const firstElement = focusableModalElements?.[0] as HTMLElement;
    const lastElement = focusableModalElements?.[
      focusableModalElements.length - 1
    ] as HTMLElement;

    if (!e.shiftKey && document.activeElement === lastElement) {
      firstElement.focus();
      return e.preventDefault();
    }

    if (e.shiftKey && document.activeElement === firstElement) {
      lastElement.focus();
      e.preventDefault();
    }
  }

  return isMounted && ref.current
    ? createPortal(
        <div
          className="c-CommonModalWrapper"
          onClick={() => handleToggle(false)}
          id={modalId}
          data-testid="common-modal-wrapper"
        >
          <div
            className={classNames("c-CommonModalWrapper__Modal", {
              "c-CommonModalWrapper__Modal_sizeMedium":
                size === ICommonModalWrapperSize.MEDIUM,
              "c-CommonModalWrapper__Modal_sizeLarge":
                size === ICommonModalWrapperSize.LARGE,
            })}
            onClick={(e) => e.stopPropagation()}
            onKeyDown={(e) => handleKeyDown(e)}
            ref={focusRef}
            data-testid="common-modal"
          >
            {children}
            <button
              className="c-CommonModalWrapper__Close"
              onClick={() => handleToggle(false)}
              data-testid="common-modal-close"
            />
          </div>
        </div>,
        ref.current,
      )
    : null;
});
