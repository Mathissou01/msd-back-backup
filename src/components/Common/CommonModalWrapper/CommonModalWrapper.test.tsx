import { fireEvent, render, screen, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import CommonModalWrapper, {
  CommonModalWrapperRef,
} from "./CommonModalWrapper";
import React from "react";

const childRef = React.createRef<CommonModalWrapperRef>();
const TestModal = (props: {
  handleClose: () => void;
  toggleValue?: boolean;
}) => (
  <>
    <main>
      <input data-testid="outside-content" />
      <button
        onClick={() => childRef.current?.toggleModal(props.toggleValue)}
      />
      <CommonModalWrapper ref={childRef} onClose={props.handleClose}>
        <input data-testid="modal-content-1" />
        <input data-testid="modal-content-2" />
        <input data-testid="modal-content-3" />
      </CommonModalWrapper>
    </main>
    <div id="modal-portal" />
  </>
);

it("renders nothing when not opened", () => {
  const handleClose = jest.fn();
  const props = { handleClose };
  const { container } = render(<TestModal {...props} />);

  const modal = screen.queryByTestId("common-modal");
  const modalContent = screen.queryByTestId("modal-content-1");
  expect(modal).not.toBeInTheDocument();
  expect(modalContent).not.toBeInTheDocument();

  expect(container).toMatchSnapshot();
});

it("renders content when opened", () => {
  const handleClose = jest.fn();
  const props = { handleClose, toggleValue: true };
  const { container } = render(<TestModal {...props} />);

  const button = screen.getByRole("button");
  fireEvent.click(button);
  const modal = screen.getByTestId("common-modal");
  const modalContent = within(modal).queryByTestId("modal-content-1");
  expect(modalContent).toBeInTheDocument();

  expect(container).toMatchSnapshot();
});

it("it traps focus for Tab and Shift+Tab navigation", async () => {
  const handleClose = jest.fn();
  const props = { handleClose };
  render(<TestModal {...props} />);
  const button = screen.getByRole("button");

  fireEvent.click(button);
  const modal = screen.getByTestId("common-modal");
  const input1 = within(modal).getByTestId("modal-content-1");
  const input2 = within(modal).getByTestId("modal-content-2");
  const input3 = within(modal).getByTestId("modal-content-3");
  const modalClose = within(modal).getByTestId("common-modal-close");
  expect(input1).toHaveFocus();

  await userEvent.tab();
  expect(input2).toHaveFocus();
  await userEvent.tab();
  expect(input3).toHaveFocus();
  await userEvent.tab();
  expect(modalClose).toHaveFocus();
  await userEvent.tab();
  expect(input1).toHaveFocus();

  await userEvent.tab({ shift: true });
  await userEvent.tab({ shift: true });
  expect(input3).toHaveFocus();
});

it("closes with button, click outside, or escape key", () => {
  const handleClose = jest.fn();
  const props = { handleClose };
  render(<TestModal {...props} />);
  const button = screen.getByRole("button");

  // First open
  fireEvent.click(button);
  const modal = screen.getByTestId("common-modal");
  const modalOpen = within(modal).queryByTestId("modal-content-1");
  expect(modalOpen).toBeInTheDocument();

  const closeCross = within(modal).getByRole("button");
  fireEvent.click(closeCross);
  expect(handleClose).toHaveBeenCalledTimes(1);
  expect(modal).not.toBeInTheDocument();

  // Second open
  fireEvent.click(button);
  const modal2 = screen.getByTestId("common-modal");
  const modalOpen2 = within(modal2).queryByTestId("modal-content-1");
  expect(modalOpen2).toBeInTheDocument();

  const clickOutside = screen.getByTestId("common-modal-wrapper");
  fireEvent.click(clickOutside);
  expect(handleClose).toHaveBeenCalledTimes(2);
  expect(modal2).not.toBeInTheDocument();

  // Third open
  fireEvent.click(button);
  const modal3 = screen.getByTestId("common-modal");
  const modalOpen3 = within(modal3).queryByTestId("modal-content-1");
  expect(modalOpen3).toBeInTheDocument();

  fireEvent.keyDown(modal3, { key: "Escape", code: "Escape" });
  expect(handleClose).toHaveBeenCalledTimes(3);
  expect(modal3).not.toBeInTheDocument();
});
