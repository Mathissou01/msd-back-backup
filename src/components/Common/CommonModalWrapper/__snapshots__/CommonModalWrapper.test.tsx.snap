// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommonModalWrapper renders content when opened 1`] = `
<div>
  <main>
    <input
      data-testid="outside-content"
    />
    <button />
  </main>
  <div
    id="modal-portal"
  >
    <div
      class="c-CommonModalWrapper"
      data-testid="common-modal-wrapper"
      id=":r1:"
    >
      <div
        class="c-CommonModalWrapper__Modal c-CommonModalWrapper__Modal_sizeMedium"
        data-testid="common-modal"
      >
        <input
          data-testid="modal-content-1"
        />
        <input
          data-testid="modal-content-2"
        />
        <input
          data-testid="modal-content-3"
        />
        <button
          class="c-CommonModalWrapper__Close"
          data-testid="common-modal-close"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`CommonModalWrapper renders nothing when not opened 1`] = `
<div>
  <main>
    <input
      data-testid="outside-content"
    />
    <button />
  </main>
  <div
    id="modal-portal"
  />
</div>
`;
